zcash_primitives/src/transaction/components/sapling/builder.rs

```diff
@@ -0,0 +1,533 @@
+                                rng.fill_bytes(&mut d);
+                                diversifier = Diversifier(d);
+                                if let Some(val) = diversifier.g_d() {
+                                    g_d = val;
+                                    break;
+                                }
+                            }
+                            (diversifier, g_d)
+                        };
+
+                        let (pk_d, payment_address) = loop {
+                            let dummy_ivk = jubjub::Fr::random(&mut rng);
+                            let pk_d = g_d * dummy_ivk;
+                            if let Some(addr) = PaymentAddress::from_parts(diversifier, pk_d) {
+                                break (pk_d, addr);
+                            }
+                        };
+
+                        let rseed = generate_random_rseed_internal(params, target_height, &mut rng);
+
+                        (
+                            payment_address,
+                            Note {
+                                g_d,
+                                pk_d,
+                                rseed,
+                                value: 0,
+                            },
+                        )
+                    };
+
+                    let esk = dummy_note.generate_or_derive_esk_internal(&mut rng);
+                    let epk = dummy_note.g_d * esk;
+
+                    let (zkproof, cv) =
+                        prover.output_proof(ctx, esk, dummy_to, dummy_note.rcm(), dummy_note.value);
+
+                    let cmu = dummy_note.cmu();
+
+                    let mut enc_ciphertext = [0u8; 580];
+                    let mut out_ciphertext = [0u8; 80];
+                    rng.fill_bytes(&mut enc_ciphertext[..]);
+                    rng.fill_bytes(&mut out_ciphertext[..]);
+
+                    // Update progress and send a notification on the channel
+                    progress += 1;
+                    if let Some(sender) = progress_notifier {
+                        // If the send fails, we should ignore the error, not crash.
+                        sender
+                            .send(Progress::new(progress, Some(total_progress)))
+                            .unwrap_or(());
+                    }
+
+                    OutputDescription {
+                        cv,
+                        cmu,
+                        ephemeral_key: epk.into(),
+                        enc_ciphertext,
+                        out_ciphertext,
+                        zkproof,
+                    }
+                }
+            })
+            .collect();
+
+        Ok((spend_descs, output_descs, tx_metadata))
+    }
+
+    pub fn create_signatures<Pr: TxProver, R: RngCore>(
+        self,
+        prover: &Pr,
+        ctx: &mut Pr::SaplingProvingContext,
+        rng: &mut R,
+        sighash_bytes: &[u8; 32],
+        tx_metadata: &SaplingMetadata,
```

Comment:
I dislike that we rely on the caller to accurately pass this between calls. I presume that in a later refactor that brings this in line with the `orchard` crate, we can persist this inside pre-authorization type metadata.